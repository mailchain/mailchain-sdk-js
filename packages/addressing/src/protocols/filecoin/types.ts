export enum FilecoinAddressType {
	SECP256K1 = 1,
	BLS12_381 = 3,
	DELEGATED = 4,
}
