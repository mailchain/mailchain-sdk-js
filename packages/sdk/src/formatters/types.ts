export type MailAddress = {
	name: string;
	address: string;
};

export type MailData = {
	id: string;
	subject: string;
	from: MailAddress;
	recipients: MailAddress[];
	carbonCopyRecipients: MailAddress[];
	blindCarbonCopyRecipients: MailAddress[];
	message: string;
	plainTextMessage: string;
};
