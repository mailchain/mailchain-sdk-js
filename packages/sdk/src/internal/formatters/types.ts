export type MailAddress = {
	name: string;
	address: string;
};

export type MailData = {
	id: string;
	subject: string;
	from: MailAddress;
	date: Date;
	recipients: MailAddress[];
	carbonCopyRecipients: MailAddress[];
	blindCarbonCopyRecipients: MailAddress[];
	message: string;
	plainTextMessage: string;
};
