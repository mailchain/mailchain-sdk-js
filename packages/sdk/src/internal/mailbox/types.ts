import { PublicKey } from '@mailchain/crypto';

export type MessagePreview = {
	mailbox: PublicKey;
	messageId: string;
	from: string;
	owner: string;
	subject: string;
	snippet: string;
	isRead: boolean;
	systemLabels: SystemMessageLabel[];
	hasAttachment: boolean;
	timestamp: Date;
	to: string[];
	cc: string[];
	bcc: string[];
};

export type Message = {
	from: string;
	replyTo?: string;
	to: string[];
	cc: string[];
	bcc: string[];
	subject: string;
	timestamp: Date;
	body: string;
};

/** Copy from cmd/inbox/internal/datastore/labels.go */
export const SYSTEM_MESSAGE_LABELS = [
	'archive',
	'draft',
	'inbox',
	'sent',
	'spam',
	'starred',
	'trash',
	'unread',
	'outbox',
] as const;
export type SystemMessageLabel = (typeof SYSTEM_MESSAGE_LABELS)[number];
export type UserMessageLabel = string;
