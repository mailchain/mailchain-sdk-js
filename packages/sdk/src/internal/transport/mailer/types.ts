import { MailAddress } from '../mail/types';

export type MailerData = {
	/**
	 * Message subject line, this will become the `Subject:` header in the email.
	 */
	subject: string;
	/**
	 * Reply-To address, this will become the `Reply-To:` header in the email.
	 */
	replyTo?: MailAddress;
	/**
	 * HTML formatted message, this will become the `text/html` part of the email.
	 */
	message: string;
	/**
	 * Plain text formatted message, this will become the `text/plain` part of the email.
	 */
	plainTextMessage: string;
};
