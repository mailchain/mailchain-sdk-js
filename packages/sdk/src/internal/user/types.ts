import { MailchainAddress, ProtocolType } from '@mailchain/addressing';
import { PublicKey } from '@mailchain/crypto';

/**
 * Represents different type of alias to be set for the {@link UserMailbox}.
 */
export type Alias = {
	address: MailchainAddress;
	isDefault: boolean;
	allowSending: boolean;
	allowReceiving: boolean;
};

/**
 * Entity for sending and receiving messages based on the provided identity key and messaging key.
 */
export type UserMailbox = {
	id: string;
	/**
	 * - `'account'` - Mailchain account (example: `'alice@mailchain.com'`).
	 * - `'wallet'` - registered mailbox based blockchain wallet identity key (example: `'0xEBaae0532dF65ee3f1623f324C9620bB84c8af8d@ethereum.mailchain.com'`)
	 */
	type: 'account' | 'wallet';
	identityKey: PublicKey;
	/** The user preferred label to shown for this mailbox. If `null`, no user preferred label is defined, the application is free to compute one at runtime. */
	label: string | null;
	/** Will contain at least one defined {@link Alias}. */
	aliases: [Alias, ...Alias[]];
	/** Ingredients for the creation of messaging private key via the keyring. */
	messagingKeyParams: {
		address: Uint8Array;
		protocol: ProtocolType;
		network: string;
		nonce: number;
	};
};
