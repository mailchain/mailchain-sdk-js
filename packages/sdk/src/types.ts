// needed for documentation
// eslint-disable-next-line @typescript-eslint/no-unused-vars, no-unused-vars
import type { ProtocolType } from '@mailchain/addressing';

/**
 * Mailchain address formatted in mail format.
 *
 * Format: `{address}@{protocol}.{network}.mailchain.com`
 * - `address` - the value of the address encoded with the protocol encoding.
 *   - Example: Ethereum address it is encoded with 0x prefixed Hex, resulting in something like `0x0E5736fbD198496Ef9A890a8D8be7538De9B2e0f`
 *   - Example: Mailchain address it is encoded with UTF-8, resulting in just regular text `alice`
 * - `protocol` - some of the supported protocols by mailchain {@link ProtocolType}
 *   - Optional: can be omitted when the protocol is 'mailchain'
 * - `network` - The network of the protocol
 *   - Optional: can be omitted if it is the main network/chain of the protocol.
 *
 * Examples:
 * - Ethereum address on Ethereum mainnet `0x0E5736fbD198496Ef9A890a8D8be7538De9B2e0f@etherum.mailchain.com`
 * - Mailchain address `alice@mailchain.com`
 */
export type Address = string;

export type Attachment = {
	/** Unique identifier for the attachment that can be used to reference it to embed it into the content of the mail content. */
	cid: string;
	/** The name to appear when showing and downloading the attachment.  */
	name: string;
	/** `Base64` encoded value of the bytes or just `Buffer` from it. */
	content: string | Buffer;
};

export type SendMailParams = {
	/** Address for sending the mail. If not owned by the authenticated user, error is thrown. */
	from: Address;
	/** Address that the replies should be sent to instead of the {@link SendMailParams.from} address. */
	replyTo?: Address;
	/** The recipients that will appear in the `To:` field. */
	to?: Address[];
	/** The recipients that will appear in the `Cc:` field. */
	cc?: Address[];
	/**
	 * The recipients that will appear in the `Bcc:` field.
	 *
	 * Note: Mailchain SDK will take care to hide these recipients from others.
	 */
	bcc?: Address[];
	/** The subject of the email. */
	subject: string;
	content: {
		/** The HTML version of the message */
		html?: string | Buffer;
		/** The plaintext version of the message */
		text?: string | Buffer;
	};
	attachments?: Attachment[];
	/** The thread that this message belongs to */
	threadId?: string;
	/**
	 * Define custom set of headers. Most messages don't need defining custom headers, but if you do need this is the way.
	 *
	 * Note: any custom header that has property defined in the standard set of the {@link SendMessageParams} will get overwritten.
	 */

	headers?: { [key: string]: string | undefined };
};

export type MessagePreview = {
	/**
	 * The Mailchain identification for message for further referencing with Mailchain.
	 *
	 * Note: Not related to the 'Message-Id' mail header.
	 */
	mailchainMessageId: string;
	from: Address;
	to: Address[];
	cc: Address[];
	bcc: Address[];
	/** The subject of the email. */
	subject: string;
	/** A short part of the message text. */
	snippet: string;
	/** List of labels applied to this message. Can contain both system labels and custom user labels. */
	labels: string[];
	/** The thread that this message belongs to */
	threadId: string;
	/** Message creation timestamp  */
	date: Date;
	/** Estimated size in bytes of the message. */
	sizeEstimate: number;
	hasAttachment: boolean;
	/** @deprecated will get removed in favor of message grouping */
	owner: string;
};

/** Attachment that part of the message. */
type IncomingAttachment = {
	/** Unique identifier for the attachment that can be used to reference it to embed it into the content of the mail content. */
	cid: string;
	/** The name to appear when showing and downloading the attachment.  */
	name: string;
	/** Estimated size in bytes of the attachment.  */
	sizeEstimate: number;
};

/** Attachment that part of the message and it is embedded as part of the content. */
type EmbeddedIncomingAttachment = IncomingAttachment & {
	content: Buffer;
};

// Self note: Separate `Message` is returned by the request in order not to do unnecessary network traffic and encryption/decryption.
/** The content for the message of {@link MessagePreview}. */
export type Message = {
	mailchainMessageId: string;
	/**
	 * Contains the content of the message.
	 *
	 * - `'html'` - The HTML content part of the message as UTF-8 encoded string.
	 * - `'text'` - The plain/text content part of the message as UTF-8 encoded string.
	 * - `'raw'` - The entire email message in an RFC 2822 formatted and base64url encoded string.
	 */
	content: {
		type: 'html' | 'text' | 'raw';
		value: string;
	};
	allAttachments: IncomingAttachment[];
	embeddedAttachments: EmbeddedIncomingAttachment[];
};
