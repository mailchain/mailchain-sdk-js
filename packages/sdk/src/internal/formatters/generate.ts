import { createMessageComposer } from '@mailchain/message-composer';
import { MailAddress, MailData } from './types';

export const createMimeMessage = async (
	mailData: MailData,
): Promise<{
	original: string;
	visibleRecipients: string;
	blindRecipients: { recipient: MailAddress; content: string }[];
}> => {
	const msg = createMessageComposer()
		.id(mailData.id)
		.date(mailData.date)
		.subject(mailData.subject)
		.from(mailData.from)
		.recipients('To', ...mailData.recipients)
		.recipients('Cc', ...mailData.carbonCopyRecipients)
		.recipients('Bcc', ...mailData.blindCarbonCopyRecipients)
		.message('html', Buffer.from(mailData.message))
		.message('plain', Buffer.from(mailData.plainTextMessage));

	if (mailData.replyTo) msg.replyTo(mailData.replyTo);

	const builtMsg = await msg.build();

	return {
		original: builtMsg.forSender,
		visibleRecipients: builtMsg.forVisibleRecipients,
		blindRecipients: builtMsg.forBlindedRecipients.map(([recipient, content]) => ({
			recipient: { name: recipient.name!, address: recipient.address },
			content,
		})),
	};
};
