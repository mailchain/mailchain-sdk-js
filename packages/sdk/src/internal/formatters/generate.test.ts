import { createMimeMessage } from './generate';
import { MailData } from './types';

describe('createMimeMessage', () => {
	const mailData: MailData = {
		id: '123@mailchain.local',
		from: { address: '1337@mailchain.com', name: '1337' },
		recipients: [
			{ address: 'rec1@mailchain.local', name: 'rec1' },
			{ address: 'rec2@mailchain.local', name: 'rec2' },
		],
		carbonCopyRecipients: [
			{ address: 'rec3@mailchain.local', name: 'rec3' },
			{ address: 'rec4@mailchain.local', name: 'rec4' },
		],
		blindCarbonCopyRecipients: [
			{ address: 'rec5@mailchain.local', name: 'rec5' },
			{ address: 'rec6@mailchain.local', name: 'rec6' },
		],
		subject: 'Subject ðŸ¤£ðŸ˜²ðŸ¥³ðŸ˜²ðŸ¥³ðŸ™‚ å¤§å¤§å¤§å¤§',
		message: ['line 1', 'line2', '', 'line4 ðŸ¤£ðŸ˜²ðŸ¥³ðŸ˜²ðŸ¥³ðŸ™‚'].join('\n'),
		plainTextMessage: ['line 1', 'line2', '', 'line4'].join('\n'),
		date: new Date('2022-06-06'),
	};

	it('should generate correct message', async () => {
		const messages = await createMimeMessage(mailData);

		expect(removeRandomBoundaries(messages.original)).toMatchSnapshot('ORIGINAL');
		expect(removeRandomBoundaries(messages.visibleRecipients)).toMatchSnapshot('VISIBLE');
		expect(removeRandomBoundaries(messages.blindRecipients[0].content)).toMatchSnapshot(
			'BLIND rec5@mailchain.local',
		);
		expect(removeRandomBoundaries(messages.blindRecipients[1].content)).toMatchSnapshot(
			'BLIND rec6@mailchain.local',
		);
	});
});

function removeRandomBoundaries(msg: string): string {
	msg = msg.replaceAll(/boundary.*/g, `boundary="boundary"`);
	msg = msg.replaceAll(/--.*/g, `--boundary`);

	return msg;
}
