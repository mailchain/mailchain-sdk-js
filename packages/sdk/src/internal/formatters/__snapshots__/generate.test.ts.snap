// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createMimeMessage should generate correct message: BLIND rec5@mailchain.local 1`] = `
"MIME-Version: 1.0
Date: Mon, 06 Jun 2022 00:00:00 +0000
Message-ID: 123@mailchain.local
Subject: =?UTF-8?B?U3ViamVjdCDwn6Sj8J+YsvCfpbPwn5iy8J+ls/CfmYIg5aSn5aSn5aSn5aSn?=
From: \\"1337\\" <1337@mailchain.com>
Reply-To: \\"7311\\" <7331@mailchain.com>
To: \\"rec1\\" <rec1@mailchain.local>,
 \\"rec2\\" <rec2@mailchain.local>
Cc: \\"rec3\\" <rec3@mailchain.local>,
 \\"rec4\\" <rec4@mailchain.local>
Bcc: \\"rec5\\" <rec5@mailchain.local>
Content-Type: multipart/alternative; boundary=\\"boundary\\"

--boundary
Content-Type: text/html; charset=\\"UTF-8\\"
Content-Transfer-Encoding: base64

bGluZSAxCmxpbmUyCgpsaW5lNCDwn6Sj8J+YsvCfpbPwn5iy8J+ls/CfmYI=
--boundary
Content-Type: text/plain; charset=\\"UTF-8\\"
Content-Transfer-Encoding: base64

bGluZSAxCmxpbmUyCgpsaW5lNA==
--boundary"
`;

exports[`createMimeMessage should generate correct message: BLIND rec6@mailchain.local 1`] = `
"MIME-Version: 1.0
Date: Mon, 06 Jun 2022 00:00:00 +0000
Message-ID: 123@mailchain.local
Subject: =?UTF-8?B?U3ViamVjdCDwn6Sj8J+YsvCfpbPwn5iy8J+ls/CfmYIg5aSn5aSn5aSn5aSn?=
From: \\"1337\\" <1337@mailchain.com>
Reply-To: \\"7311\\" <7331@mailchain.com>
To: \\"rec1\\" <rec1@mailchain.local>,
 \\"rec2\\" <rec2@mailchain.local>
Cc: \\"rec3\\" <rec3@mailchain.local>,
 \\"rec4\\" <rec4@mailchain.local>
Bcc: \\"rec6\\" <rec6@mailchain.local>
Content-Type: multipart/alternative; boundary=\\"boundary\\"

--boundary
Content-Type: text/html; charset=\\"UTF-8\\"
Content-Transfer-Encoding: base64

bGluZSAxCmxpbmUyCgpsaW5lNCDwn6Sj8J+YsvCfpbPwn5iy8J+ls/CfmYI=
--boundary
Content-Type: text/plain; charset=\\"UTF-8\\"
Content-Transfer-Encoding: base64

bGluZSAxCmxpbmUyCgpsaW5lNA==
--boundary"
`;

exports[`createMimeMessage should generate correct message: ORIGINAL 1`] = `
"MIME-Version: 1.0
Date: Mon, 06 Jun 2022 00:00:00 +0000
Message-ID: 123@mailchain.local
Subject: =?UTF-8?B?U3ViamVjdCDwn6Sj8J+YsvCfpbPwn5iy8J+ls/CfmYIg5aSn5aSn5aSn5aSn?=
From: \\"1337\\" <1337@mailchain.com>
Reply-To: \\"7311\\" <7331@mailchain.com>
To: \\"rec1\\" <rec1@mailchain.local>,
 \\"rec2\\" <rec2@mailchain.local>
Cc: \\"rec3\\" <rec3@mailchain.local>,
 \\"rec4\\" <rec4@mailchain.local>
Bcc: \\"rec5\\" <rec5@mailchain.local>,
 \\"rec6\\" <rec6@mailchain.local>
Content-Type: multipart/alternative; boundary=\\"boundary\\"

--boundary
Content-Type: text/html; charset=\\"UTF-8\\"
Content-Transfer-Encoding: base64

bGluZSAxCmxpbmUyCgpsaW5lNCDwn6Sj8J+YsvCfpbPwn5iy8J+ls/CfmYI=
--boundary
Content-Type: text/plain; charset=\\"UTF-8\\"
Content-Transfer-Encoding: base64

bGluZSAxCmxpbmUyCgpsaW5lNA==
--boundary"
`;

exports[`createMimeMessage should generate correct message: VISIBLE 1`] = `
"MIME-Version: 1.0
Date: Mon, 06 Jun 2022 00:00:00 +0000
Message-ID: 123@mailchain.local
Subject: =?UTF-8?B?U3ViamVjdCDwn6Sj8J+YsvCfpbPwn5iy8J+ls/CfmYIg5aSn5aSn5aSn5aSn?=
From: \\"1337\\" <1337@mailchain.com>
Reply-To: \\"7311\\" <7331@mailchain.com>
To: \\"rec1\\" <rec1@mailchain.local>,
 \\"rec2\\" <rec2@mailchain.local>
Cc: \\"rec3\\" <rec3@mailchain.local>,
 \\"rec4\\" <rec4@mailchain.local>
Content-Type: multipart/alternative; boundary=\\"boundary\\"

--boundary
Content-Type: text/html; charset=\\"UTF-8\\"
Content-Transfer-Encoding: base64

bGluZSAxCmxpbmUyCgpsaW5lNCDwn6Sj8J+YsvCfpbPwn5iy8J+ls/CfmYI=
--boundary
Content-Type: text/plain; charset=\\"UTF-8\\"
Content-Transfer-Encoding: base64

bGluZSAxCmxpbmUyCgpsaW5lNA==
--boundary"
`;
